crates/prover/src/adapter/instructions.rs

```diff
@@ -0,0 +1,723 @@
+pub struct StatesByOpcodes {
+    pub store_add_fp_fp: Vec<StateData>, // [fp + off2] = [fp + off0] + [fp + off1]
+    pub store_add_fp_imm: Vec<StateData>, // [fp + off2] = [fp + off0] + imm
+    pub store_sub_fp_fp: Vec<StateData>, // [fp + off2] = [fp + off0] - [fp + off1]
+    pub store_sub_fp_imm: Vec<StateData>, // [fp + off2] = [fp + off0] - imm
+    pub store_deref_fp: Vec<StateData>,  // [fp + off2] = [fp + off0]
+    pub store_double_deref_fp: Vec<StateData>, // [fp + off2] = [[fp + off0] + off1]
+    pub store_imm: Vec<StateData>,       // [fp + off2] = imm
+    pub store_mul_fp_fp: Vec<StateData>, // [fp + off2] = [fp + off0] * [fp + off1]
+    pub store_mul_fp_imm: Vec<StateData>, // [fp + off2] = [fp + off0] * imm
+    pub store_div_fp_fp: Vec<StateData>, // [fp + off2] = [fp + off0] / [fp + off1]
+    pub store_div_fp_imm: Vec<StateData>, // [fp + off2] = [fp + off0] / imm
+    pub call_abs_fp: Vec<StateData>,     // call abs [fp + off0]
+    pub call_abs_imm: Vec<StateData>,    // call abs imm
+    pub call_rel_fp: Vec<StateData>,     // call rel [fp + off0]
+    pub call_rel_imm: Vec<StateData>,    // call rel imm
+    pub ret: Vec<StateData>,             // ret
+    pub jmp_abs_add_fp_fp: Vec<StateData>, // jmp abs [fp + off0] + [fp + off1]
+    pub jmp_abs_add_fp_imm: Vec<StateData>, // jmp abs [fp + off0] + imm
+    pub jmp_abs_deref_fp: Vec<StateData>, // jmp abs [fp + off0]
+    pub jmp_abs_double_deref_fp: Vec<StateData>, // jmp abs [[fp + off0] + off1]
+    pub jmp_abs_imm: Vec<StateData>,     // jmp abs imm
+    pub jmp_abs_mul_fp_fp: Vec<StateData>, // jmp abs [fp + off0] * [fp + off1]
+    pub jmp_abs_mul_fp_imm: Vec<StateData>, // jmp abs [fp + off0] * imm
+    pub jmp_rel_add_fp_fp: Vec<StateData>, // jmp rel [fp + off0] + [fp + off1]
+    pub jmp_rel_add_fp_imm: Vec<StateData>, // jmp rel [fp + off0] + imm
+    pub jmp_rel_deref_fp: Vec<StateData>, // jmp rel [fp + off0]
+    pub jmp_rel_double_deref_fp: Vec<StateData>, // jmp rel [[fp + off0] + off1]
+    pub jmp_rel_imm: Vec<StateData>,     // jmp rel imm
+    pub jmp_rel_mul_fp_fp: Vec<StateData>, // jmp rel [fp + off0] * [fp + off1]
+    pub jmp_rel_mul_fp_imm: Vec<StateData>, // jmp rel [fp + off0] * imm
+    pub jnz_fp_fp: Vec<StateData>,       // jmp rel [fp + off1] if [fp + off0] != 0
+    pub jnz_fp_imm: Vec<StateData>,
+}
+
+#[derive(Debug, Default)]
+pub struct Instructions {
+    pub initial_registers: VmRegisters,
+    pub final_registers: VmRegisters,
+    pub states_by_opcodes: StatesByOpcodes,
+}
+
+impl Instructions {
+    pub fn push_instr<I>(
+        &mut self,
+        mut memory: I,
+        registers: VmRegisters,
+        clock: u32,
+        memory_cache: &mut MemoryCache,
+    ) -> Result<(), InstructionError>
+    where
+        I: Iterator<Item = MemoryEntry>,
+    {
+        let mut state_data = StateData {
+            registers,
+            memory_args: Default::default(),
+        };
+
+        let instruction = memory
```

Comment:
how do you know that the instruction is located at the next value in memory?
i thought the instruction depended on memory[pc]?

Reply:
apart from that I think it'd be nice to have a Opcode struct that has a From<QM31>, such that you can jsut do instruction.value.into().opcode_id()